src/repeated_game.jl

```diff
@@ -438,3 +440,118 @@ function outerapproximation(rpd::RepGame2; nH=32, tol=1e-8, maxiter=500,
+               plib=getlibraryfor(2, T), tol::Float64=1e-5, u=nothing)
+
+    # Initialize W0 with each entries of payoff bimatrix
+    v_old = _payoff_points(rpd.sg)
+
+    if u == nothing
+        u = [minimum(rpd.sg.players[1].payoff_array),
+             minimum(rpd.sg.players[2].payoff_array)]
+    end
+
+    # create VRepresentation and Polyhedron
+    V = SimpleVRepresentation(v_old)
+    p = polyhedron(V, plib)
+    H = SimpleHRepresentation(p)
+    # get rid of redundant vertices
+    V = removevredundancy(V, H)
+    p = polyhedron(V, plib)
+
+    # calculate the best deviation gains
+    # normalize with (1-delta)/delta
+    best_dev_gains1, best_dev_gains2 = (1-rpd.delta)/rpd.delta .* _best_dev_gains(rpd.sg)
+
+    for iter = 1:max_iter
+
+        v_new = Vector{T}(0) # to store new vertices
+        # step 1
+        for a2 in 1:rpd.sg.nums_actions[2]
+            for a1 in 1:rpd.sg.nums_actions[1]
+                payoff1 = rpd.sg.players[1].payoff_array[a1, a2]
+                payoff2 = rpd.sg.players[2].payoff_array[a2, a1]
+                IC1 = u[1] + best_dev_gains1[a1, a2]
+                IC2 = u[2] + best_dev_gains2[a2, a1]
+
+                # check if the payoff point is interior
+                # first check if it satisifies IC
+                if all([payoff1, payoff2] .> [IC1, IC2])
+                    # then check if it is in the polyhedron
+                    if [payoff1, payoff2] in H
+                        push!(v_new, payoff1, payoff2)
+                    end
+                end
+
+                # find out the intersections of polyhedron and IC boundaries
+                p_IC = polyhedron(SimpleHRepresentation(-eye(2), -[IC1, IC2]), plib)
+                p_inter = intersect(p_IC, p)
+                V = SimpleVRepresentation(p_inter)
+                for i = 1:nvreps(V)
+                    if V.V[i, 1] ≈ IC1 || V.V[i, 2] ≈ IC2
+                        push!(v_new, (rpd.delta * V.V[i, :] +
+                                      (1 - rpd.delta) * [payoff1, payoff2])...)
+                    end
+                end
+            end
+        end
+
+        v_new = reshape(v_new, 2, :)'
+
+        # get rid of redundant points
+        V = SimpleVRepresentation(v_new)
+        p = polyhedron(V, plib)
+        H = SimpleHRepresentation(p)
+        V = removevredundancy(V, H)
+        p = polyhedron(V, plib)
+
+        # check if it's converged
+        # first check if the numbers of vertices are the same
+        if size(v_new) == size(v_old)
+            # then check the euclidean distance
+            if norm(v_new-v_old) < tol
+                println("converged in $(iter) iterations")
+                break
+            end
+        end
+
+        # check if max_iter is reached
+        if iter == max_iter
+            warn("Maximum Iteration Reached")
+        end
+
+        v_old = v_new
+
+        # step 2
+        # update u
+        u_ = [minimum(v_new[:, 1]) minimum(v_new[:, 2])]
```

Comment:
Given that `u` is always one dimensional, I'd put a comma here to make `u_` one dimensional as well.

Below when you write `u = u_` you are changing `u` between a 1 dimensional array and 2 dimensional array (this might also be a source of instability). See below for example

```
julia> u = [rand(), rand()]
2-element Array{Float64,1}:
 0.0212796
 0.144139 

julia> u_ = [rand() rand()]
1×2 Array{Float64,2}:
 0.8921  0.408161

julia> if all(u_ .> u)
           u = u_
       end
1×2 Array{Float64,2}:
 0.8921  0.408161
julia> u
1×2 Array{Float64,2}:
 0.8921  0.408161
```

Reply:
You are right. At the beginning I set both `u` and `u_` to be two dimensional, but then I change `u` to one dimensional, while forget to change `u_` as well. Thanks for finding this out!